### PR TITLE
Deploy tests

### LINKED ISSUE #49: Contract deploy tests
https://github.com/software-mansion/protostar/issues/2108

Deploy is already implemented in new runner (`cheatcodes_hint_processor.rs`). Tests verifying if it actually works should be added. If not, it should be fixed in this or follow-up issue.

Test cases to cover apart from basic cases:
- [ ] Check if `class_hash` exists, is correct and declared (and what happens if non-existing one does get passed)
- [ ] ~~Check if `contract_address` is correct and aligned with the calculated one (and what hapens if incorrect one gets passeed~~) Contract address no longer provided in `deploy`
- [ ] Check if the constructor is invoked
- [x] Check what happens when constructor panics

### PR DESCRIPTION
**Pull request by Arcticae**
Closes #49 

## Introduced changes

Added test cases:
- Verifying if constructor is invoked
- Passing incorrect `class_hash` to deploy
- Added a macro for checking the output of the test

## Breaking changes

None

## Checklist
- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code

### CODE REVIEW

File: starknet-foundry/crates/forge/tests/integration/deploy.rs [hunk 1]
```diff
@@ -223,3 +223,132 @@ fn test_deploy_fails_on_too_many_constructor_arguments() {
...
+        use starknet::ContractAddress;
+        use starknet::Felt252TryIntoContractAddress;
+        
+        #[starknet::interface]
+        trait ThingGetter<TContractState> {
+            fn get_thing(self: @TContractState) -> felt252;
+        }
+        
+        #[test]
+        fn deploy_not_existing_class_hash() {
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/integration/deploy.rs:294]: Nit but the test name could be changed

**Review by cptartur — APPROVED** [review]
Like the macro for asserting outputs

File: starknet-foundry/crates/forge/tests/integration/common/runner.rs [hunk 2]
```diff
@@ -164,3 +164,30 @@ macro_rules! assert_failed {
...
     }};
 }
+
+#[macro_export]
+macro_rules! assert_case_output_contains {
+    ($result:expr, $test_case_name:expr, $asserted_msg:expr) => {{
+        use forge::test_case_summary::TestCaseSummary;
+
+        let test_case_name = $test_case_name;
+        let test_name_suffix = format!("::{test_case_name}");
```
Comment (piotmag769) [inline starknet-foundry/crates/forge/tests/integration/common/runner.rs:174]: It is a bit error prone (two test functions can have the same name), but it is the best we can come up with for now I think

File: starknet-foundry/crates/forge/tests/integration/deploy.rs [hunk 3]
```diff
@@ -223,3 +223,132 @@ fn test_deploy_fails_on_too_many_constructor_arguments() {
...
+    let test = test_case!(
+        indoc!(
+            r#"
+        use result::ResultTrait;
+        use cheatcodes::RevertedTransactionTrait;
+        use cheatcodes::PreparedContract;
+        use array::ArrayTrait;
+
+        #[test]
+        fn deploy_not_existing_class_hash() {
```
Comment (piotmag769) [inline starknet-foundry/crates/forge/tests/integration/deploy.rs:238]: [nit]
```suggestion
        fn deploy_non_existing_class_hash() {
```

File: starknet-foundry/crates/forge/tests/integration/deploy.rs [hunk 4]
```diff
@@ -223,3 +223,132 @@ fn test_deploy_fails_on_too_many_constructor_arguments() {
...
+        
+        #[starknet::interface]
+        trait ThingGetter<TContractState> {
+            fn get_thing(self: @TContractState) -> felt252;
+        }
+        
+        #[test]
+        fn deploy_invokes_constructor() {
+            let mut calldata = ArrayTrait::new();
+            calldata.append(420);
```
Comment (piotmag769) [inline starknet-foundry/crates/forge/tests/integration/deploy.rs:296]: 420

**Review by piotmag769 — APPROVED** [review]

File: starknet-foundry/crates/forge/tests/integration/common/runner.rs (hunk 2 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/common/runner.rs:174]: you can do `module::name` and it will still work

**Pull request merged**